test/queue.test.js

```diff
@@ -0,0 +1,248 @@
+'use strict'
+const { test } = require('tap')
+const { once, when, immediate, timeout } = require('./util')
+const { PassThrough } = require('stream')
+const queue = require('../lib/queue')
+
+test('add / shift', async ({ is, same }) => {
+  const q = queue()
+
+  q.add({ port: 8000, host: '127.0.0.1' })
+
+  const { peer } = q.shift()
+
+  same(peer, { port: 8000, host: '127.0.0.1' })
+  is(q.shift(), null)
+})
+
+test('requeue', async ({ is }) => {
+  const q = queue({
+    requeue: [ 100, 200, 300 ]
+  })
+
+  q.add({ port: 8080, host: '127.0.0.1' })
+
+  const info = q.shift()
+
+  is(q.requeue(info), true)
+  await once(q, 'readable')
+  is(q.shift(), info)
+  is(q.requeue(info), true)
+  await once(q, 'readable')
+  is(q.shift(), info)
+  is(q.requeue(info), true)
+  await once(q, 'readable')
+  is(q.shift(), info)
+  is(q.requeue(info), false)
```

Comment:
If you remove the info now from the queue completely using q.remove(info) after re-adding it it should reset the queuing state

Reply:
This reminds me that we prob want to remove the peer from the queue when a requeue fails in swarm.js. I don't think we do that atm. That measn that if the dht rediscovers it later it'll be able to requeue it.

Reply:
documenting discussion follow-up: we need to hold onto a a peer for a while after requeue fails before full removal